dill/settings.py

```diff
@@ -21,5 +23,177 @@
+                else:
+                    filter = 'type:%s' % name
+                module_filters.add(filter, rule_type=rule_type)
+        if '%s.funcs' % rule in section:
+            for code in section['%s.funcs' % rule].strip('\n').splitlines():
+                if code.startswith('(') and code.endswith(')'):
+                    code = code[1:-1]
+                globals_ = {}
+                if not code.startswith('lambda'):
+                    name = code.partition('(')[0]
+                    _resolve(name)  # import required modules
+                    top_level = name.partition('.')[0]
+                    globals_[top_level] = __import__(top_level)
+                filter = eval(code, globals_)
+                if not callable(filter):
+                    raise TypeError("filter is not callable: %r" % filter)
+                module_filters.add(filter, rule_type=rule_type)
+
+def read_settings(filename) -> None:
+    """Read dill settings from an INI file.
+
+    Update the ``dill.settings`` dictionary with the contents of the INI file
+    ``filename``.  Accepted file sections:
+
+      - `dill`: general :py:mod:`dill` settings
+      - `dill.module`: settings for :py:func:`dill.dump_module`
+      - `filters`: default exclude/include filters for :py:func:`dill.dump_module`
+      - `filters.<module_name>`: module-specific filters for
+        :py:func:`dill.dump_module`, where `<module_name>` is the complete module
+        path in the form `module[.submodule...]`
+
+    Accepted option values for general settings:
+
+      - boolean options (case insensitive): yes, no, on, off, true, false
+      - `protocol`: DEFAULT_PROTOCOL, HIGHEST_PROTOCOL, 0, 1, 2, 3, ...
+      - `fmode`: HANDLE_FMODE, 0, CONTENTS_FMODE, 1, FILE_FMODE, 2
+
+    .. IMPORTANT: The demo config file below is used in test_settings.py
+    .. Lexer 'pacmanconf' generates better highlighting than 'ini'.
+    .. code-block:: pacmanconf
+
+        [dill]
+        # General settings
+        ## Stored in dill.settings.
+        protocol = HIGHEST_PROTOCOL
+        byref = yes
+
+        [dill.session]
+        # Settings for dill.dump_module()
+        ## Stored in dill.session.settings.
+        refonfail = no
+
+        [filters]
+        # Default exclude/include filters for dill.dump_module()
+        ## Stored in dill.settings['dump_module']['filters'].
+        exclude.names = some_var, SomeClass
+        exclude.regexes = '_.+'
+        exclude.types = function, ModuleType, io.BytesIO
+        exclude.funcs =
+            lambda obj: type(obj.value) == int
+            dill.session.size_filter('10 KB')
+        include = _keep_this_var, '__.+__'
+
+        [filters.some.module]
+        # Filter rules specific to the module 'some.module'
+        ## Reuse regex filters defined in the previous section.
+        ## Option 'include' is unset, will fall back to default 'include' filters.
```

Comment:
confusing. a lot of this documentation is a bit confusing.  maybe this needs to be rethought on how to present it

Reply:
the format itself seems ok and fairly standard at first pass

Reply:
I assume the `#` are comments, and the `##` have no special meaning.

Reply:
> The format itself seems ok and fairly standard at first pass. It's basically standard for the `configparser`, is that correct? If so, it doesn't need much more explanation other than (1) what format it is using, and (2) what the fields and possible settings are.

Yes. About the format, I think it's sufficient to document the very basic syntax and refer to the official `configparser` documentation for details.

> Do we want the level of docs in the config file docs in `docs/source/conf.py`, or should we rely on docs elsewhere (i.e. not in-file) to document the file?

I'm not sure I understood it well but, as we were discussing in the other thread, the complete documentation of the config file format and more complex examples could go in a standalone sphinx docs page, linked in the `read_settings()` docstring.

> I assume the `#` are comments, and the `##` have no special meaning.

That's correct. I'm thinking of substituting the `# Section header` and `## comment` format by `# Section header #` and `# comment` with an empty line between them.